src/Common/FileSegment.cpp

```diff
@@ -162,6 +162,18 @@ String FileSegment::getDownloader() const
     return downloader_id;
 }
 
+String FileSegment::getQueryId() const
+{
+    std::lock_guard segment_lock(mutex);
+
+    auto npos = downloader_id.find(":");
+
+    if (npos == std::string::npos)
+        return {};
+    else
+        return downloader_id.substr(0, npos);
+}
```

Comment:
I'll answer here to your message:
> Sorry for not being clear, what I meant was that we may evict a file_segment that is not belong to the current query in tryReserveForMainList(). The query_context we actually want to update is the one that holds the file_segment (I think it's the file_segment's downloader). But it may not be easy to find the query_context from file_segment, because each file_segment will clear its downloader_id after being downloaded. Therefore, I can only judge whether the cache in a query_context has been swapped out in tryReserveForQuery().

so 

> The query_context we actually want to update is the one that holds the file_segment (I think it's the file_segment's downloader)

but file segment can be hold by multiple queries, so we need a list of query id's. 

> But it may not be easy to find the query_context from file_segment, because each file_segment will clear its downloader_id after being downloaded.

True, and it is also reset after each `CachedReadBufferFromRemoteFS::nextImpl` call. And it only includes `DOWNLOADING` state, but we need any state. 

I am not sure whether there is a good solution as any solution will either noticeably increase memory usage or have extra time complexity. So looks like my suggestion which resulted in the need to find out query context from file segment is not good. I will think a bit more about this, may be I will come up with some idea.